edk2toollib/uefi/edk2/path_utilities.py

```diff
@@ -19,64 +19,88 @@
+        #   - self.PackagePathList: List[str]
+        #   - self.WorkspacePath: str
+
+        self.PackagePathList = []
+        self.WorkspacePath = ""
+
+        workspace_candidate_path = Path(ws)
+
+        if not workspace_candidate_path.is_absolute():
+            workspace_candidate_path = Path(os.getcwd(), ws)
+
+        if not workspace_candidate_path.is_dir():
+            raise NotADirectoryError(
+                errno.ENOENT,
+                os.strerror(errno.ENOENT),
+                workspace_candidate_path.resolve())
+
+        self.WorkspacePath = str(workspace_candidate_path)
+
+        candidate_package_path_list = []
+        for a in package_path_list:
+            if os.path.isabs(a):
+                candidate_package_path_list.append(Path(a))
             else:
-                # see if workspace relative
-                wsr = os.path.join(ws, a)
-                if (os.path.isdir(wsr)):
-                    self.PackagePathList.append(wsr)
+                wsr = Path(ws, a)
+                if wsr.is_dir():
+                    candidate_package_path_list.append(wsr)
                 else:
                     # assume current working dir relative.  Will catch invalid dir when checking whole list
-                    self.PackagePathList.append(os.path.abspath(os.path.join(os.getcwd(), a)))
-
-        error = False
-        for a in self.PackagePathList[:]:
-            if (not os.path.isdir(a)):
-                self.logger.log(logging.ERROR if error_on_invalid_pp else logging.WARNING,
-                                "Invalid package path entry {0}".format(a))
-                self.PackagePathList.remove(a)  # remove invalid path
-                error = True
-
-        # report error
-        if (error and error_on_invalid_pp):
-            raise Exception("Invalid package path directory(s)")
-
-        # for each package path, trace from packagepath to the
-        # either the workspace root or filesystem root and verify
-        # no *.DEC file exists. This would signify a nested package.
-        for package_path in self.PackagePathList:
-            p = Path(package_path)
-            ws = self.WorkspacePath
-            while p != p.parent:
-                if str(p).lower() == str(ws).lower():
-                    break
-                if len(glob.glob(f'{p}/*dec')) != 0:
-                    raise Exception(f'Nested packages not allowed. Pkg path [{package_path}] nested in Package [{p}]')
-                p = p.parent
+                    candidate_package_path_list.append(Path(os.getcwd(), a))
+
+        for a in candidate_package_path_list:
+            if not a.is_dir():
+                if error_on_invalid_pp:
+                    raise NotADirectoryError(
+                        errno.ENOENT,
+                        os.strerror(errno.ENOENT),
+                        a.resolve())
+                else:
+                    return
```

Comment:
I agree we don't want to keep the invalid package path, but there are still other valid package paths in the packagepathlist that should be kept. Your code removes all of them. I think you're confusing the user even more by removing all package paths rather than only the bad one(s). 

To me it seems like the option of not raising an error on an invalid package path has the use case of allowing a user to add package paths that may sometimes be valid and may sometimes not be valid without having to do the actual filtering (the filtering is done by Ed2Path).

What you're insinuating here is that we don't want to do that for them, and if that's the case (which I'm not against), then I don't see a legitimate reason to keep `error_on_invalid_pp` at all.

Reply:
#113 Is why we have error_on_invalid_pp. Does your change to Pr_Eval resolve this so we can remove error_on_invalid_pp?

Reply:
The goal of the change was:

1. Inform the user in an obvious and documented way their package path is invalid - the exception
2. When the user does something wrong be consistent - always prevent them from continuing by setting zero package paths

This is simply a different philosophical approach to handling invalid user input. I'm more in favor of well-defined input expectations and preventing special handling (and complicated interfaces) from creeping in for exceptional cases. This often leads to technical debt across layers of functionality as it kind of is here now. I agree that with this approach `error_on_invalid_pp` should go away.

Given #113 explicitly allowed this behavior, I will update the code to continue to support it. Modifying this is not a major goal of this change.

Reply:
Behavior brought back in PR update https://github.com/tianocore/edk2-pytool-library/pull/186/commits/31b16aa53b23925c6ef929798e38fe31b0834cd6